### DISCUSSION TITLE
"Spinner" while EntityAction is running?

### DISCUSSION THREAD

**Discussion by sturlath** [Megabit/Blazorise#2852] [Q&A]
Is it possible to kick off the `LoadingTemplate` when you have gone through a action (clicked yes on a modal popup)?

Say you have
1. List of items in a DataGrid
2. You select the Delete EntityAction  (calls the delate API and gets back new list)
3. You get the ConfirmationMessage, press Yes
4. The modal closes 
5. No UI indication something is happening 
6. Suddenly list of items, minus the deleted just pop on to the grid.

I'm going about this all wrong?

**Comment by David-Moreira** [comment]
Can you show a sample code?
If you are using ReadData Loading should come up whenever it is called. Don't forget to set the LoadingTemplate.
If you want to control the Loading Flag you have two options:
- inherit DataGrid and control it yourself.
![image](https://user-images.githubusercontent.com/22283161/131748289-768a04c5-0760-463f-bd77-34682a09ce65.png)
- wrap DataGrid on a component or page where you have your own loading flag, and make the DataGrid or Loading display accordingly.

**Reply by sturlath** [reply]
Sorry I just forgot to answer this back (got hold up in personal matters) so here goes.

Here is my code with one action that askes if you want to delete the entry, if so the DeleteEventAsync is called.
The LoadingTemplate is never invoked except the first time when OnReadData is called.

```xml
<DataGrid TItem="EventDto"
      Data="EventList"
      ReadData="@OnReadData"
      TotalItems="TotalCount"
      ShowPager="true"
      Responsive="true"
      PageSize="PageSize">
    <ChildContent>
        <DataGridEntityActionsColumn TItem="EventDto" @ref="@EntityActionsColumn">
            <DisplayTemplate>
                <EntityActions TItem="EventDto" EntityActionsColumn="@EntityActionsColumn">
                    <EntityAction TItem="EventDto"
                              RequiredPolicy="Delete"
                              Clicked="() => DeleteEventAsync(context)"
                              ConfirmationMessage="@(()=> "Are you sure?")"
                              Text=""Delete"">
                    </EntityAction>
                </EntityActions>
            </DisplayTemplate>
        </DataGridEntityActionsColumn>
        <DataGridColumn TItem="EventDto"
                    Field="StreamEvent.Name"
                    Caption=""Name">
        </DataGridColumn>
    </ChildContent>
    <EmptyTemplate>
        <div class="box">
            Not found!
        </div>
    </EmptyTemplate>
    <LoadingTemplate>
        <SpinKit Type="SpinKitType.Wave" />
    </LoadingTemplate>
</DataGrid>
```

and then the C# code

```csharp
  private IReadOnlyList<EventDto> EventList{ get; set; }

  private async Task DeleteEventAsync(EventWithNavigationPropertiesDto input)
  {
      await EventsAppService.DeleteAsync(input.Event.Id);

      EventList = await GetEventsThroughApiCodeNotRelevantAsync();
      await InvokeAsync(StateHasChanged);
  }
```

When I get the data in my OnReadData() method it shows the spinner but when I call the delete method (after pressing confirm) I never call the OnReadData() method again. 

Can I invoke call to  OnReadData() from my delete method?

**Reply by David-Moreira** [reply]
Hello, 
Make a ref to `Datagrid` and call `Reload`

**Reply by sturlath** [reply]
I just saw this https://github.com/Megabit/Blazorise/issues/1620 issue and it talks about using dataGridRef.Reload()
so now I´m trying to add `@ref="grid"` on `<DataGrid> `but without luck. I haven't been using the @ref stuff so I´ll keep on trying but this might be what I'm looking for.

**Reply by sturlath** [reply]
Ok got it.. will add code in 2 min!

**Reply by sturlath** [reply]
Add this to your DataGrid properties `@ref="@YourGrid"`
Add this to your C# `private DataGrid<EventDto> YourGrid{ get; set; }`
Call this at the top of your "DeleteAsync" method `await YourGrid.Reload();`

**Reply by sturlath** [reply]
Sorry this is **not what I need**.. 

If I place this at the top of the delete method it will 

1. reload the data and show the spinner (this could be very fast.. spinner on and then off)
2. **then** I delete the entry (this could take a long time and **with no spinner**)
3. Then flicker the an update when the list is updated 

So this is **not** what I need.

I basically need something to spin while the C# DeleteAsync() is executed

Just saw your reply @David-Moreira sorry for spamming you but like you can see above I don't think that will work.

**Reply by David-Moreira** [reply]
I was answering this.
> Can I invoke call to OnReadData() from my delete method?

Reload invokes it behind the scenes. For your use case might make sense to call it after the delete call so you refresh the data. 
As for calling loading while deleting, only way is to inherit and set it yourself.

We might take a look at enabling an API to set loading manually @strski agree?

**Reply by stsrki** [reply]
Adding a new API seems like the most logical way. DataGrid shows the loading template only when the actual data is loading. And in this case, only after delete it will trigger the load.

But even if we add the new API it still doesn't make sense. Deleting isn't loading. Maybe better is to have a DeletingTemplate that could be triggered once the Delete command is activated.

**Reply by sturlath** [reply]
I kinda like the `DeletingTemplate` trail of thought, but since that is not available is there a way for me to add a spinner after I press confirm on the modal? 

Would be nice If the modal could go blurry with activity indicator on top while its action would finish.

**Reply by David-Moreira** [reply]
@stsrki I kinda half agree with having the DeletingTemplate, if we do introduce it, we also need an EdittingTemplate and a "Newing"Template. Or a more generic one, for when the Datagrid is "comitting" stuff or "doing work". 

Maybe there could be a more generic one that users could use for whatever they want? Like a "BusyTemplate"? And they'd have access to the `SetBusy` method? 

@sturlath if you want to use the `LoadingTemplate` from the DataGrid, you have to override the Datagrid.
For the project I am working on, I just overrode the Datagrid.
Something like 
```
    public class MyDataGrid<TItem> : DataGrid<TItem>
    {
        protected override Task OnParametersSetAsync()
        {
            base.RowRemoving = EventCallback.Factory.Create<CancellableRowChange<TItem>>( this, MyRowRemoving );
            return base.OnParametersSetAsync();
        }

        private async Task MyRowRemoving( CancellableRowChange<TItem> cancellableRowChange )
        {
            IsLoading = true;
            await InvokeAsync( StateHasChanged );

            //Do my deletion work
            //.....

            //Cancel if needed // You can even cancel if you don't want the Datagrid to mutate the internal collection, but still delete the record from your storage.
            //cancellableRowChange.Cancel = true;

            IsLoading = false;
            await InvokeAsync( StateHasChanged );
        }
    }
```
Now on the modal itself, I don't think there's an easy way. But you can always hack away at it! Set up some dummy column. Put the spinner in there, and just manage whenever you want it to show up or not.

**Reply by stsrki** [reply]
I like the idea of BusyTemplate. Or SavingTemplate maybe?

**Reply by sturlath** [reply]
ActiveTemplate/IsActiveTemplate/WhileActiveTemplate? Btw thanks @David-Moreira I´ll let you know what I do and how it went when I get down to it.

**Reply by David-Moreira** [reply]
I personally don't like those "Active" templates in the context of the Datagrid. 

I think the active definition would not be quite clear from just looking at those. What I mean is, I don't think a regular user could quickly infer from the naming you just suggested when are each one of those applicable.

I liked both the Saving and busy templates. While the SavingTemplate already has everything setup for the user ready to use, the other one gives more flexibility. 

Maybe I would go so far as to have only the BusyTemplate, and whenever Datagrid does an operation other then reading data it could set itself to busy. 
If user wants to control himself then he can use the SetBusy api. Of course the template will only render/ be used if set, otherwise everything works as is. 


Alright, let us know how it goes. :)

**Reply by stsrki** [reply]
Yes, BusyTemplate is the clearest on what the intent is and what it might be used for. And it will take care of most of the scenarios.
